### DISCUSSION TITLE
Storage Drivers

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12295] [Ideas]
This has been discussed on various internal trackers and work has been started, converting to a public issue now.

# Goals
**Primary**: Make the cache storage of the Dagger Engine pluggable and configurable, including both local and remote cache
* Layer caching - Users should be able to get the same seamless remote caching enabled by Dagger Cloud but with their own storage locations (as opposed to the ones created by Dagger's infra). 
   * The underlying goal being that users who can't and don't prefer to use our infra for one reason or another are still able to benefit from the same feature set provided by Dagger Cloud caching.
* Cache volumes - should be able to be stored on configurable directories from the host system of the dagger engine, with any synchronization managed in userspace
   * E.g. it should be easy to use external systems like GHA cache directories, K8s persistent storage backends like EBS/Ceph/etc. and similar to synchronize these cache volumes based on what makes sense in the users' infra

**Secondary**: Make caching more configurable, observable and debuggable. These are very important, but are more like improvements we need to make while implementing the above goals, which is why I'm calling them secondary for now.

Caching is a well established _Hard Problem™️_ but this is made worse currently by the fact that it's hard to know what's being stored in the cache, what's being pruned when+why, etc.

Users (including dagger's own developers) should easily be able to
* Know what is being stored in cache currently (what operation on which engine created it) and associated metadata like the size, the number of uses, etc.
* Prune the cache (either by configuring settings for automatic garbage-collection or through manual pruning API calls)
* Understand *why* cache misses are happening.
   * This is **_by far_** the trickiest + most subtle problem of all but also the most common question, so worth the effort. 
   * It involves understanding which _previous_ operation a user expected to get a cache hit on with a _new_ operation (already tricky) *and then* breaking down each operation into all the inputs to cache key to provide a "diff" that breaks down where the cache keys diverged.

# Design
## Cache Volumes
This work has started and is in a POC working state in [this PR](https://github.com/dagger/dagger/pull/7646).

Important design points (all of which are still open for discussion):
1. _By default_, engines will have the same behavior as today where cache volumes are stored in the same local state as local layer cache, get automatically pruned in the background by the engine etc.
2. But, engines can now be configured to instead store cache volumes in a separate arbitrary directory inside the engine container
   * This directory can be bind-mounted into the engine container from arbitrary directories on the host.
3. The way the directory on the host is configured is still being worked on, but
   * At bare minimum, this should be a setting available in the `/etc/dagger/engine.toml` file
   * It would be convenient to also support as an env var set on the engine container
   * Also worth considering would be to support this in the CLI as an official flag, which would be applied during engine provisioning
      * This in particular gets into compute-driver related concerns, so it *may* make sense to hold off and address once we get to that project
4. In terms of the "userspace" API for cache volumes, the goal is to have no changes there right now.
    * The idea being that the underlying storage mechanism and user API for using cache volumes should remain decoupled and orthogonal
5. Cache volumes that are shared between concurrently running engines should be safe to use
    * e.g. if engines are running in k8s and using a Ceph persistent storage volume as the backend, it should be safe to have the same volume attached to multiple engines
    * Implementation-wise, this is currently relying on flock support being available in the underlying filesystem storing the cache volumes.
    * Synchronization on shared filesystems is notoriously difficult, so it may be worth pulling back on this for now if flock turns out to be problematic and full of race conditions; will require testing.
6. Cache volumes stored on the host should _**not**_ be automatically pruned and instead be pruned by manual API calls made by the user
   * [Recent work](https://github.com/dagger/dagger/pull/7767) added some provisional APIs for cache state and pruning, which will be expanded to support this.
   * The goal is for there to be a CLI frontend UX for controlling this, though it is all in the core API so it could be programmed by users through an SDK too.

## Layer Cache
Work has not started here yet, but design has been discussed. Main points so far:
1. Users should be able to store cache in their self-provisioned storage backends (S3/GCS/Azure/R2/etc. bucket, container registry) rather than have to use Dagger Cloud's storage backends
   * Ideally, this mechanism should be fully pluggable and perhaps even implemented as a module, but that's more aspirational at this point and not a primary goal yet.
2. We should retain the behavior of Dagger Cloud caching where engines transparently sync cache in the background rather than being tied to individual clients synchronously importing/exporting cache
3.. Coordination between engines is the hard part, especially in regards to pruning. High-level options
   * We come up with a clever way for distributed engines to do this decentralized (i.e. algorithms for safely merging/compacting/pruning cache manifests)
   * We give users a way to manage centralized infra that helps distributed engines coordinate this (i.e. they have to provision a shared DB like any sql db, nosql ones like dynamo, etc.)
   * Dagger cloud helps with the coordination; it just no longer *stores* any cache.
   * Some combination of the above (they aren't inherently mutually exclusive)

There's a few more ideas floating around that need to be fleshed out but are worth mentioning:
1. This is the area most in need of improvements in terms of observability and debuggability (e.g. "why am I getting cache misses"). This needs to be an important input consideration every step of the way.
2. The current design has mostly be in regards to remote cache storage, but there could be benefits to giving local cache storage the same properties of remote cache storage in terms of shareability.
   * E.g. right now multiple engines can't run off of the same local cache state. If they *could* though, that opens up a lot more architectures for the engine as a whole (i.e. each client could have its own dedicated engine while still sharing underlying local cache)
   * The big idea here is essentially that local cache could be "just another cache backend" rather than a special case. It's only difference from remote cache would be that it comes first and is preferred over remote cache if there's a match in both (cache tiering)
3. The idea of fundamentally reworking how caching works in Dagger has come up a lot recently. The idea being that we may want to detach from Buildkit's implementation and replace it with something "dagql-native" (whereas right now we essentially have two caching systems between dagql and buildkit)
    * This would be a large effort, so while it may be very beneficial in terms of overall simplicity and ease of adding new features, it needs to be approached with caution.

# Work so far
- [x] https://github.com/dagger/dagger/pull/7767 - Adds initial APIs for querying and controlling local cache, base for the rest of the work in terms of API-ifying cache configuration/control/observability.
- [ ] https://github.com/dagger/dagger/pull/7646  - Support for storing cache volumes on the host

**Comment by JacobLey** [comment]
As an extension to configurable caching, I envision support for tiered caching as well.

For example, a remote cache (like S3 or Dagger Cloud), as well as local/secondary caches.

For example my organization can have one large shared remote cache (something like Dagger itself, or S3 would _naively_ work well).
Only CI jobs would have the ability to _write_ to this cache, everyone else gets readonly.

Then individuals still have their own writable cache, perhaps a separately maintained "private" S3 account, or even just their local machine (or both)

So for any given cacheable action (pseudo-code):

```
// Some string computed on all the inputs, like container data + the exec text
inputHash = engine.computeHash(input)

// Check for pre-computed result in cache, favoring lower tier (assume lower = more local = cheaper/faster to download)
for (cache in caches.reverse()) {
   outputResult = cache.getResult(inputHash)
   if (outputResult != null) {
      // Attempt to re-write the cache to even lower "local" tier, so future cache lookups are faster
      for (lowerCache in caches.reverse()) {
        if (lowerCache.isLowerTier(cache) && lowerCache.isWriteable()) {
          lowerCache.write(inputHash, outputResult);
          break;
        }
      }
      return outputResult
   }
}
// cache miss
outputResult = engine.compute(input);

// Write output to all caches
for (cache in caches) {
   if (cache.isWriteable()) {
      cache.write(inputHash, outputResult);
   }
}
```

The result is that once code is pushed to CI, every member of the team gets immediate benefits of pre-computed containers.

But when individuals are testing locally (lots of junk and failure), they don't pollute the shared cache, which may be limited in storage size, or at the very least not be free (especially when compared to running locally)

it also means if some cache starts pruning data (e.g. a lifecycle rule that deletes data that hasn't been accessed in >1 month), local caches can just pick up the slack.

It could also allow organizations to share their builds publicly.

For example, the Dagger organization itself should grant everyone access to Dagger's internal cache for building this repo, so anyone contributing for the first time immediately is up to speed. But when they start messing around and changing code, there is no risk to Dagger that junk/malicious builds will end up in their cache.

**Comment by syrm** [comment]
any update ?
